File: sdk/typescript/src/agent-conversions.ts
import {
  create,
  type MessageInitShape,
} from "@bufbuild/protobuf";

import {
  AgentOutputSchema,
  AgentStructuredOutputSchema,
  AgentTextOutputSchema,
  AgentMessagePartSchema,
  AgentMessagePartType as ProtoAgentMessagePartType,
  AgentMessageSchema,
  AgentTurnDisplaySchema,
  AgentTurnStructuredOutputSchema,
  AgentTurnTextOutputSchema,
  type AgentMessage as ProtoAgentMessage,
  type AgentMessagePart as ProtoAgentMessagePart,
  type AgentOutput as ProtoAgentOutput,
  type AgentTurn as ProtoAgentTurn,
  type AgentTurnDisplay as ProtoAgentTurnDisplay,
} from "./internal/gen/v1/agent_pb.ts";
import {
  SubjectContextSchema,
  AgentToolRefSchema,
  type SubjectContext as ProtoSubjectContext,
  type AgentToolRef as ProtoAgentToolRef,
} from "./internal/gen/v1/app_pb.ts";
import {
  jsonFromValue,
  valueFromJson,
  type JsonInput,
} from "./protocol.ts";
import {
  optionalObjectFromStruct,
  optionalStruct,
} from "./protocol-internal.ts";
import type {
  AgentOutput,
  AgentMessage,
  AgentMessagePart,
  AgentMessagePartType,
  AgentToolRef,
  AgentTurnOutput,
  AgentTurnDisplay,
} from "./agent.ts";
import type { Subject, SubjectInput } from "./api.ts";

export function agentTurnDisplayFromProto(
  display?: ProtoAgentTurnDisplay | undefined,
): AgentTurnDisplay | undefined {
  if (display === undefined) {
    return undefined;
  }
  return {
    kind: display.kind,
    phase: display.phase,
    text: display.text,
    label: display.label,
    ref: display.ref,
    parentRef: display.parentRef,
    input: display.input === undefined ? undefined : jsonFromValue(display.input) as JsonInput,
    output: display.output === undefined ? undefined : jsonFromValue(display.output) as JsonInput,
    error: display.error === undefined ? undefined : jsonFromValue(display.error) as JsonInput,
    action: display.action,
    format: display.format,
    language: display.language,
  };
}

export function agentTurnDisplayToProto(
  display: AgentTurnDisplay | undefined,
): MessageInitShape<typeof AgentTurnDisplaySchema> | undefined {
  if (!display) {
    return undefined;
  }
  return {
    kind: display.kind ?? "",
    phase: display.phase ?? "",
    text: display.text ?? "",
    label: display.label ?? "",
    ref: display.ref ?? "",
    parentRef: display.parentRef ?? "",
    input: display.input === undefined ? undefined : valueFromJson(display.input),
    output: display.output === undefined ? undefined : valueFromJson(display.output),
    error: display.error === undefined ? undefined : valueFromJson(display.error),
    action: display.action ?? "",
    format: display.format ?? "",
    language: display.language ?? "",
  };
}

export function agentOutputFromProto(
  output?: ProtoAgentOutput | undefined,
): AgentOutput | undefined {
  switch (output?.kind.case) {
    case "text":
      return { text: {} };
    case "structured":
      if (output.kind.value.schema === undefined) {
        throw new Error("output.structured.schema is required");
      }
      return {
        structured: {
          schema: optionalObjectFromStruct(output.kind.value.schema) ?? {},
        },
      };
    default:
      return undefined;
  }
}

export function agentOutputToProto(
  output: AgentOutput | undefined,
): MessageInitShape<typeof AgentOutputSchema> | undefined {
  if (output === undefined) {
    throw new Error("agent output is required");
  }
  const textSet = output.text !== undefined;
  const structuredSet = output.structured !== undefined;
  if (textSet === structuredSet) {
    throw new Error("exactly one of output.text or output.structured is required");
  }
  if (textSet) {
    return {
      kind: {
        case: "text",
        value: create(AgentTextOutputSchema, {}),
      },
    };
  }
  if (structuredSet) {
    if (output.structured.schema === undefined) {
      throw new Error("output.structured.schema is required");
    }
    return {
      kind: {
        case: "structured",
        value: create(AgentStructuredOutputSchema, {
          schema: optionalStruct(output.structured.schema),
        }),
      },
    };
  }
  throw new Error("exactly one of output.text or output.structured is required");
}

export function agentTurnOutputFromProto(
  output: ProtoAgentTurn["output"],
): AgentTurnOutput | undefined {
  switch (output.case) {
    case "text":
      return { text: output.value.text };
    case "structured":
      return {
        structured: {
          text: output.value.text,
          value: optionalObjectFromStruct(output.value.value),
        },
      };
    default:
      return undefined;
  }
}

export function agentTurnOutputToProto(
  output: AgentTurnOutput | undefined,
): ProtoAgentTurn["output"] {
  if (output === undefined) {
    return { case: undefined };
  }
  const textSet = output.text !== undefined;
  const structuredSet = output.structured !== undefined;
  if (textSet === structuredSet) {
    throw new Error("exactly one of output.text or output.structured is required");
  }
  if (textSet) {
    return {
      case: "text",
      value: create(AgentTurnTextOutputSchema, {
        text: output.text ?? "",
      }),
    };
  }
  if (structuredSet) {
    return {
      case: "structured",
      value: create(AgentTurnStructuredOutputSchema, {
        text: output.structured.text ?? "",
        value: optionalStruct(output.structured.value),
      }),
    };
  }
  throw new Error("exactly one of output.text or output.structured is required");
}

export function agentMessageFromProto(message: ProtoAgentMessage): AgentMessage {
  return {
    role: message.role,
    text: message.text,
    parts: message.parts.map(agentMessagePartFromProto),
    metadata: optionalObjectFromStruct(message.metadata),
  };
}

export function agentMessageToProto(
  message: AgentMessage,
): MessageInitShape<typeof AgentMessageSchema> {
  return {
    role: message.role ?? "",
    text: message.text ?? "",
    parts: message.parts?.map(agentMessagePartToProto) ?? [],
    metadata: optionalStruct(message.metadata),
  };
}

export function agentMessagePartFromProto(
  part: ProtoAgentMessagePart,
): AgentMessagePart {
  return {
    type: part.type as AgentMessagePartType,
    text: part.text,
    json: optionalObjectFromStruct(part.json),
    toolCall: part.toolCall === undefined ? undefined : {
      id: part.toolCall.id,
      toolId: part.toolCall.toolId,
      arguments: optionalObjectFromStruct(part.toolCall.arguments),
    },
    toolResult: part.toolResult === undefined ? undefined : {
      toolCallId: part.toolResult.toolCallId,
      status: part.toolResult.status,
      content: part.toolResult.content,
      output: optionalObjectFromStruct(part.toolResult.output),
    },
    imageRef: part.imageRef === undefined ? undefined : {
      uri: part.imageRef.uri,
      mimeType: part.imageRef.mimeType,
    },
  };
}

export function agentMessagePartToProto(
  part: AgentMessagePart,
): MessageInitShape<typeof AgentMessagePartSchema> {
  return {
    type: part.type ?? ProtoAgentMessagePartType.UNSPECIFIED,
    text: part.text ?? "",
    json: optionalStruct(part.json),
    toolCall: part.toolCall === undefined ? undefined : {
      id: part.toolCall.id ?? "",
      toolId: part.toolCall.toolId ?? "",
      arguments: optionalStruct(part.toolCall.arguments),
    },
    toolResult: part.toolResult === undefined ? undefined : {
      toolCallId: part.toolResult.toolCallId ?? "",
      status: part.toolResult.status ?? 0,
      content: part.toolResult.content ?? "",
      output: optionalStruct(part.toolResult.output),
    },
    imageRef: part.imageRef === undefined ? undefined : {
      uri: part.imageRef.uri ?? "",
      mimeType: part.imageRef.mimeType ?? "",
    },
  };
}

export function agentToolRefFromProto(ref: ProtoAgentToolRef): AgentToolRef {
  return {
    app: ref.app,
    operation: ref.operation,
    connection: ref.connection,
    instance: ref.instance,
    title: ref.title,
    description: ref.description,
    credentialMode: ref.credentialMode,
    system: ref.system,
    runAs: agentRunAsSubjectFromProto(ref.runAs),
  };
}

export function agentToolRefToProto(ref: AgentToolRef): ProtoAgentToolRef {
  return create(AgentToolRefSchema, {
    app: ref.app ?? "",
    operation: ref.operation ?? "",
    connection: ref.connection ?? "",
    instance: ref.instance ?? "",
    title: ref.title ?? "",
    description: ref.description ?? "",
    credentialMode: ref.credentialMode ?? "",
    system: ref.system ?? "",
    runAs: agentRunAsSubjectToProto(ref.runAs),
  });
}

function agentRunAsSubjectFromProto(
  subject?: ProtoSubjectContext | undefined,
): Subject | undefined {
  if (subject === undefined) {
    return undefined;
  }
  return {
    id: subject.id,
    credentialSubjectId: subject.credentialSubjectId,
    email: subject.email,
  };
}

function agentRunAsSubjectToProto(
  subject?: SubjectInput | undefined,
): ProtoSubjectContext | undefined {
  if (subject === undefined) {
    return undefined;
  }
  return create(SubjectContextSchema, {
    id: subject.id ?? "",
    credentialSubjectId: subject.credentialSubjectId ?? "",
    email: subject.email ?? "",
  });
}

File: sdk/go/agent_conversions.go
package gestalt

import (
	"fmt"

	proto "github.com/valon-technologies/gestalt/server/rpc/protov1/v1"
)

func subjectFromProto(value *proto.SubjectContext) *Subject {
	if value == nil {
		return nil
	}
	return &Subject{
		ID:                  value.GetId(),
		CredentialSubjectID: value.GetCredentialSubjectId(),
		Email:               value.GetEmail(),
		DisplayName:         value.GetDisplayName(),
		Scopes:              cloneStrings(value.GetScopes()),
		Permissions:         subjectPermissionsFromProto(value.GetPermissions()),
	}
}

func subjectToProto(value *Subject) *proto.SubjectContext {
	if value == nil {
		return nil
	}
	return &proto.SubjectContext{
		Id:                  value.ID,
		CredentialSubjectId: value.CredentialSubjectID,
		Email:               value.Email,
		DisplayName:         value.DisplayName,
		Scopes:              cloneStrings(value.Scopes),
		Permissions:         subjectPermissionsToProto(value.Permissions),
	}
}

func agentToolRefFromProto(value *proto.AgentToolRef) AgentToolRef {
	if value == nil {
		return AgentToolRef{}
	}
	return AgentToolRef{
		App:            value.GetApp(),
		Operation:      value.GetOperation(),
		Connection:     value.GetConnection(),
		Instance:       value.GetInstance(),
		Title:          value.GetTitle(),
		Description:    value.GetDescription(),
		CredentialMode: value.GetCredentialMode(),
		System:         value.GetSystem(),
		RunAs:          subjectFromProto(value.GetRunAs()),
	}
}

func agentToolRefsFromProto(values []*proto.AgentToolRef) []AgentToolRef {
	if len(values) == 0 {
		return nil
	}
	out := make([]AgentToolRef, 0, len(values))
	for _, value := range values {
		out = append(out, agentToolRefFromProto(value))
	}
	return out
}

func agentToolRefToProto(value AgentToolRef) *proto.AgentToolRef {
	return &proto.AgentToolRef{
		App:            value.App,
		Operation:      value.Operation,
		Connection:     value.Connection,
		Instance:       value.Instance,
		Title:          value.Title,
		Description:    value.Description,
		CredentialMode: value.CredentialMode,
		System:         value.System,
		RunAs:          subjectToProto(value.RunAs),
	}
}

func agentToolRefsToProto(values []AgentToolRef) []*proto.AgentToolRef {
	if len(values) == 0 {
		return nil
	}
	out := make([]*proto.AgentToolRef, 0, len(values))
	for _, value := range values {
		out = append(out, agentToolRefToProto(value))
	}
	return out
}

func agentOutputFromProto(value *proto.AgentOutput) *AgentOutput {
	if value == nil || value.GetKind() == nil {
		return nil
	}
	if value.GetText() != nil {
		return &AgentOutput{Text: &AgentTextOutput{}}
	}
	if structured := value.GetStructured(); structured != nil {
		return &AgentOutput{
			Structured: &AgentStructuredOutput{
				Schema: mapFromStruct(structured.GetSchema()),
			},
		}
	}
	return nil
}

func agentOutputToProto(value *AgentOutput) (*proto.AgentOutput, error) {
	if value == nil {
		return nil, nil
	}
	switch {
	case value.Text != nil && value.Structured != nil:
		return nil, fmt.Errorf("agent output cannot set both text and structured")
	case value.Text != nil:
		return &proto.AgentOutput{Kind: &proto.AgentOutput_Text{Text: &proto.AgentTextOutput{}}}, nil
	case value.Structured != nil:
		schema, err := structFromAny(value.Structured.Schema)
		if err != nil {
			return nil, err
		}
		return &proto.AgentOutput{
			Kind: &proto.AgentOutput_Structured{
				Structured: &proto.AgentStructuredOutput{Schema: schema},
			},
		}, nil
	default:
		return nil, nil
	}
}

func cloneAgentStringMap(value map[string]string) map[string]string {
	if len(value) == 0 {
		return nil
	}
	out := make(map[string]string, len(value))
	for key, item := range value {
		out[key] = item
	}
	return out
}
